### PR TITLE
Rounding rules for limits and derived amounts

### LINKED ISSUE #812: Rounding rules for limits and derived amounts
This rules apply to the calculation of baselines, limits and derived values: CalcProd, CalcCons, ProdArt5, Prod/Cons of LabUse, Prod/Imp of EssenUse, etc.

By default, all values are rounded to 1 decimal, with the exception of annex group C/I for (reporting period >= 2011 or special cases).

Special cases:
- reporting period 2009: parties CD, CG, DZ, EC, ER, GQ, GW, HT, LC, MA, MK, MZ, NE, NG, SZ, FJ, PK, PH
- reporting period 2010: parties DZ, EC, ER, HT, LC, LY, MA, NG, PE, SZ, TR, YE, FJ, PK, PH

When the baseline or limit refers to C/I and a period listed above, they must also have 2 decimals,

### PR DESCRIPTION
**Pull request by catalinjitea**
closes #812

### CODE REVIEW

File: ozone/core/models/aggregation.py [hunk 1]
```diff
@@ -281,12 +306,34 @@ def calculate_totals(self):
...
+        ]
+
     def save(self, *args, **kwargs):
         """
         At each save, we need to recalculate the totals.
         """
         self.calculate_totals()
 
+        for field_name in self.get_quantity_fields():
+            if getattr(self, field_name) is not None and getattr(self, field_name) != '':
```
Comment (melish) [inline ozone/core/models/aggregation.py]: Se repetă de trei ori `getattr(self, field_name)`

File: ozone/core/management/commands/mkfixtures.py [hunk 2]
```diff
@@ -485,7 +485,12 @@ def baseline_map(self, f, row):
                 'name',
                 party_type + 'Prod'
             )
-            f['baseline'] = row['CalcProd']
+            # Don't need to call get_decimals for 'BaseA5' and 'BaseNA5' periods
+            # because is not a special case and we will round to 2 decimal places.
+            baseline_prod = row['CalcProd']
+            if baseline_prod:
+                baseline_prod = round(row['CalcProd'], 1)
```
Comment (sorinstelian) [inline ozone/core/management/commands/mkfixtures.py]: Comment says we will round to 2 decimal places but we round to 1.

File: ozone/core/management/commands/mkfixtures.py [hunk 3]
```diff
@@ -570,7 +578,10 @@ def calc_avg(self, data, periods):
...
             s += data[period]['ProdArt5']
-        baseline = round(s / len(periods), 2)
+        baseline = s / len(periods)
         if data['prod_transfers']:
             baseline += data['prod_transfers']
-        return baseline
+
+        # Periods will always be between '1995' and '2000'
+        # so we round to 2 decimal places.
+        return round(baseline, 1)
```
Comment (sorinstelian) [inline ozone/core/management/commands/mkfixtures.py]: Comment says we will round to 2 decimal places but we round to 1.

File: ozone/core/management/commands/mkfixtures.py (hunk 2 above)
Comment (catalinjitea) [reply ozone/core/management/commands/mkfixtures.py]: Comments were wrong.

**Comment by melish** [comment]
Fix rounding for these BDN limits:
image

**Comment by melish** [comment]
The rounding rules should also be applied to the API (/api/submissions/{submission_id}/aggregations/)

**Comment by catalinjitea** [comment]
> Fix rounding for these BDN limits:
> 
> image

Fixed, except for Netherlands/2000/AI. To be discussed.

**Comment by catalinjitea** [comment]
> The rounding rules should also be applied to the API (/api/submissions/{submission_id}/aggregations/)

Done in a0852b5.

File: ozone/core/models/aggregation.py [hunk 4]
```diff
@@ -268,6 +294,31 @@ def calculate_totals(self):
...
+        """
+        return [
+            'baseline_prod',
+            'baseline_cons',
+            'baseline_bdn',
+            'limit_prod',
+            'limit_cons',
+            'limit_bdn',
+            'calculated_production',
+            'calculated_consumption',
```
Comment (melish) [inline ozone/core/models/aggregation.py]: Probably more fields and additional rounding rules should be added here to preserve the legacy calculation rules, but we'll do that in another issue, after clarifications.

File: ozone/core/management/commands/mkfixtures_limits.py [hunk 5]
```diff
@@ -66,7 +69,7 @@ def handle(self, *args, **options):
                                 continue
                         else:
                             baseline = baseline.baseline
-                        limit = baseline * cm.allowed
+                        limit = (100 * cm.allowed * baseline) / 100
```
Comment (melish) [inline ozone/core/management/commands/mkfixtures_limits.py]: Probably useless, but doesn't hurt

File: ozone/core/management/commands/mkfixtures_limits.py [hunk 6]
```diff
@@ -100,13 +103,10 @@ def handle(self, *args, **options):
...
                         days2 = (period.end_date - cm2.start_date).days + 1
-                        limit = round(
-                            (
-                                baseline1 * cm1.allowed * days1
-                                + baseline2 * cm2.allowed * days2
-                            ) / ((period.end_date - period.start_date).days + 1),
-                            2
-                        )
+                        limit = (
+                            (100 * cm1.allowed * days1 * baseline1) / 100
```
Comment (melish) [inline ozone/core/management/commands/mkfixtures_limits.py]: Same here. Multiply + divide by 100 doesn't change anything

**Pull request merged**